Fix linkJdk task

Xerus2000 opened the pull request:
I am currently not able to build the project due to https://github.com/gradle/gradle/issues/9985, so this PR provides a fix. Furthermore, I think it does not make sense to scan the location as an output directory, because it is not of interest to this task.

sghpjuikit commented:
Got stuck on this yesterday. I'll merge tonight.

Btw I also got stuck on kotlinc task. Didnt figure it out yet. It is severe as it can not be worked around and prevents build.

sghpjuikit commented:
Id appreciate it if you wrote what the actual problem with the linkJdk task was, because I do not know if your solution can fix the issue nor test it.

I did confirm that changing the linkLocation will not run the task the way it was before, so I suggest adding the Input annotation.

Apart from that your handling of the location property seems to me to be wrong.

Xerus2000 commented on buildSrc/src/main/kotlin/GradleUtil.kt:
```diff
@@ -26,26 +26,26 @@ val String.sysProp: String?
+         Files.createSymbolicLink(linkLocation.toPath(), jdkPath)
       } catch (e: Exception) {
-         println("Couldn't create a symbolic link from $linkDir to $jdkPath: $e")
+         logger.warn("Couldn't create a symbolic link at $linkLocation to $jdkPath: $e")
```
You can run grade with --info and --debug to see the respective messages, this prevents polluting the log with trivial messages. It is Gradles logger mechanism, and you should be aware of the advantages of such a thing already ;)

sghpjuikit replied:
I wasnt aware.

Running with --info is a bit of a hassle (especially from IDE). Also the prints are only present if the task actually runs so I feel like that accomplishes the job fairly well, but alright... let's use that.

I dont think this 'consistency' will prove to have added value though - because you do not usually use trace/debug and suddenly you lose the info about what the task is doing - this is even more palpable if you are first timer cloning the project.

Xerus2000 replied:
If you "are first timer cloning the project" then you don't want random messages showing up. If a task completes successfully without any problems, it should stay silent. That is the norm in Gradle.

sghpjuikit replied:
I guess I just want to always run with INFO then.

Xerus2000 commented on buildSrc/src/main/kotlin/GradleUtil.kt:
```diff
@@ -26,26 +26,26 @@ val String.sysProp: String?
-   /** Directory/link that will contain the JDK. Should be inside the project. */
-   @OutputDirectory lateinit var linkDir: File
+   /** Location of the link to the JDK. */
+   @Input @Internal lateinit var linkLocation: File
```
The problem is that @OutputDirectory scans the contents if the whole directory for changes, which is absolutely bonkers, and as long as it is there my issue will probably persist. You could do a separate up-to-date check that checks if the link points to the right location.

Xerus2000 replied:
We don't need to always run the task, as said, we can add a custom up-to-date-check. I'll look into it.

Xerus2000 replied:
done

sghpjuikit replied:
Well, or that.

Xerus2000 commented:
1. Custom onlyIf will now check if there is already a link pointing to the right target
2. I have changed gradle-wrapper.properties to use all instead of bin, so that we can see the sources in IDEA
3. I have removed the jdkVersion property, because we don't care whether the java version changes, we only care if the java.home has moved.

Xerus2000 commented:
I'd say ready to merge? I would suggest a regular merge, but you could also squash.

sghpjuikit commented:
1 The onlyIf seems much better - easy to understand and predict too
2 ok

sghpjuikit commented:
@Xerus2000 

1
Unfortunately your linkJdk upToDate check is broken. It does not handle FileNotFoundException. Caused me some pain just now.
I noticed you only catch certain exceptions, but I was too busy to comment on it. Was a bug in the end.

2
Also, the idea to use log.info() instead of println() is terrible. If linkJDK task succeeds, I do not know what JDK was actually linked! That is very impractical.
I do not agree that if all gradle tasks succeed, there should be no output. Why do you want the build to be black box and force developer to investigate afterwards what actually happened and whether his settings were applied?
Gradle is not running with --info by default )an doing so does not appear to be a general practice) so log.info() is useless as far as I can see.

3
And exceptions are also not logged. When linkJdk check failed due to the above bug, I got:`* What went wrong: Could not evaluate onlyIf predicate for task ':linkJdk'.`
So much information at once! Why do I need to run --debug when the error information is already there. When it fails, give me the problem asap!
So I added to the catch that rethrows: `logger.error("Failed linkJdk task up to date check: $e")` which provides all the necessary info.

4
When you set up linkJdk originally to use jdk.home, I was concerned, but didnt realize what was wrong. Until today.

Link jdk should use `org.gradle.java.home` as a link target, not `java.home`, as the former is project specific. I do not even wish to use `java.home` as a fallback (though `org.gradle.java.home` probably falls back to that on auto). Why would anyone make their projects depend on single system wide property???

Anyway, I am unable to use other JDK than 12, because my property to use specific JDK is ignored. I'm fixing that now.

Xerus2000 commented:
1) Oh, that was an oversight ^^

2) Gradle tries to stay really concise. Most people don't need some output to know which JDK was linked, it just happens and works because they probably only have one. If you want it differently, I think you can change the default gradle parameters in IDEA settings, but we should adhere to the standards.

3) As for exceptions: It only gives brief information if something goes wrong. If you are anticipating an error, just run it with `--stacktrace`, you could even set that as default as well, that will give you the full exception.

sghpjuikit commented:
Yeah I too dont wish to go against conventions, it just seems bizzare to me that even when the error happens, it is not logged/displayed and gradle wants me to do a 2nd run.
